context("Test annotation and gtf: annotation.r; gff.r; go-analysis.R; ngsReferenceFiles.r")

param = ezParam()
gtfFile = system.file("extdata/genes.gtf", package="ezRun", mustWork=TRUE)
param$ezRef@refFeatureFile = system.file("extdata/genes.gtf", package="ezRun", mustWork=TRUE)
param$ezRef@refAnnotationFile = "anno.txt"
param$ezRef@refChromSizesFile = "example.txt"
fp = "/srv/GT/reference/Saccharomyces_cerevisiae/Ensembl/EF4/Sequence/WholeGenomeFasta/genome.fa"
param$ezRef@refFastaFile = fp
seqAnno = writeAnnotationFromGtf(param)
gtf = ezLoadFeatures(param, gtfFile)

test_that("Tests functions in annotation.r", {
  featureAnno = ezFeatureAnnotation(param, rownames(seqAnno), "gene")
  expect_is(seqAnno, "data.frame")
  expect_identical(names(seqAnno), names(featureAnno))
  geneMapping = getGeneMapping(param, seqAnno)
  expect_is(geneMapping, "array")
  hasMapping = hasGeneMapping(param, seqAnno)
  expect_true(hasMapping)
})

test_that("Tests basic functions in gff.r", {
  expect_is(gtf, "data.frame")
  transcripts = ezGffAttributeField(gtf$attributes, field="transcript_id", attrsep="; *", valuesep=" ")
  expect_is(transcripts, "character")
  expect_identical(nrow(gtf), length(transcripts))
  newAttribute = ezBuildAttributeField(gtf[ , c("gene_id", "gene_name", "transcript_id", "tss_id")],"gtf")
  expect_is(newAttribute, "character")
  expect_identical(nrow(gtf), length(newAttribute))
  gtf2 = ezReadGff(system.file("extdata/genes.gtf", package="ezRun", mustWork=TRUE))
  expect_is(gtf2, "data.frame")
  trimmedTranscripts = gffTrimTranscripts(gtf)
  expect_is(trimmedTranscripts, "data.frame")
  ranges = gffToRanges(gtf)
  expect_is(ranges, "GRanges")
  groupedGtf = groupGff(gtf)
  expect_is(groupedGtf, "data.frame")
  expect_lt(nrow(groupedGtf), nrow(gtf))
  groupedRanges = gffGroupToRanges(gtf, grouping = gtf$Parent)
  expect_is(groupedRanges, "GRanges")
  expect_lt(length(groupedRanges), length(ranges))
})

test_that("Tests more functions from gff.r", {
  transcriptAnnotation = transcriptAnnoFromGtf(gtf)
  expect_is(transcriptAnnotation, "data.frame")
  moreTranscripts = addTranscriptsToGffExons(gtf)
  expect_gt(nrow(moreTranscripts), nrow(gtf))
  exonNumber = getExonNumber(gtf)
  expect_is(exonNumber, "integer")
  expect_identical(length(exonNumber), nrow(gtf))
  transcriptDB = ezTranscriptDbFromRef(param$ezRef)
  expect_is(transcriptDB, "TxDb")
  tr2Gene = getTranscript2Gene(gtf)
  expect_is(tr2Gene, "character")
  counts = countIsoformsPerGene(tr2Gene)
  expect_is(counts, "table")
  transcriptSequences = getTranscriptSequences(param)
  expect_is(transcriptSequences, "DNAStringSet")
  gcAndWidth = getTranscriptGcAndWidth(param)
  expect_is(gcAndWidth$gc, "numeric")
  expect_is(gcAndWidth$featWidth, "integer")
  ensemblTypes = getEnsemblTypes(gtf)
  expect_is(ensemblTypes, "character")
})

test_that("Tests annotation functions related to GO (from go-analysis.R)", {
  do = doGo(param, seqAnno)
  expect_is(do, "logical")
  has = hasGoAnnotation(seqAnno)
  expect_is(has, "logical")
  expect_null(getGOparents("noGoTerm"))
  parents = getGOparents("GO:0034767")
  expect_is(parents, "character")
  expect_true(all(grepl("GO", names(parents))))
  added = addGoParents(c("GO:0034767", "GO:0034768"), "BP")
  expect_is(added, "list")
  expect_true(all(grepl("GO", unlist(added))))
  goList = goStringsToList(c("GO:0034762; GO:0034763", "GO:0034764; GO:0034765"))
  expect_is(goList, "list")
  expect_equal(length(unlist(goList)), 4)
})

test_that("Tests cleanGenomeFiles() from ngsReferenceFiles.r", {
  cleanedGenome = cleanGenomeFiles(param$ezRef@refFastaFile, gtfFile)
  expect_is(cleanedGenome$genomeSeq, "DNAStringSet")
  expect_is(cleanedGenome$gtf, "data.frame")
})

ezSystem("rm -fr anno.txt")
