
# p2886
setwd("/scratch/gtan/debug/CellRanger-p2886")
library(ezRun)
param = list()
param[['cores']] = '8'
param[['ram']] = '40'
param[['scratch']] = '100'
param[['node']] = ''
param[['process_mode']] = 'SAMPLE'
param[['name']] = 'CellRangerCount'
param[['refBuild']] = 'Homo_sapiens/Ensembl/GRCh38.p10/Annotation/Release_91-2018-02-26'
param[['refFeatureFile']] = 'genes.gtf'
param[['specialOptions']] = ''
param[['mail']] = 'ge.tan@fgcz.ethz.ch'
param[['dataRoot']] = '/srv/gstore/projects'
param[['resultDir']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12'
output = list()
output[['Name']] = 'B9P9_NPCHP11_NDd5'
output[['ResultDir [File]']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12/B9P9_NPCHP11_NDd5'
output[['Report [Link]']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12/B9P9_NPCHP11_NDd5/outs/web_summary.html'
output[['BAM [File]']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12/B9P9_NPCHP11_NDd5/outs/possorted_genome_bam.bam'
output[['BAI [File]']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12/B9P9_NPCHP11_NDd5/outs/possorted_genome_bam.bam.bai'
output[['Species']] = 'Homo sapiens (human)'
param[['refBuild']] = 'Homo_sapiens/Ensembl/GRCh38.p10/Annotation/Release_91-2018-02-26'
param[['refFeatureFile']] = 'genes.gtf'
output[['CountMatrix [File]']] = 'p2886/CellRangerCount_29556_NOV19_o4775_2018-09-16--08-18-12/B9P9_NPCHP11_NDd5/outs/filtered_gene_bc_matrices/10X_Ref_Human_GRCh38.p10_20180305_Release_91/matrix.mtx'
output[['Sample Id [B-Fabric]']] = 'bfs_184126'
output[['Order Id [B-Fabric]']] = '4775'
input = list()
input[['Name']] = 'B9P9_NPCHP11_NDd5'
input[['Condition']] = ''
input[['Species']] = 'Homo sapiens (human)'
input[['FragmentSize']] = '0'
input[['SampleConc']] = '1'
input[['Tube']] = 'p2886_4775/2'
input[['Sample Id']] = 'bfs_184126'
input[['PlatePosition']] = 'B1'
input[['LibConc_100_800bp']] = '76.6'
input[['LibConc_qPCR']] = '0'
input[['RIN']] = '0'
input[['Adapter1']] = 'NA'
input[['Adapter2']] = 'NA'
input[['strandMode']] = 'NA'
input[['LibraryPrepKit']] = '10X Genomics Single Cell'
input[['EnrichmentMethod']] = 'None'
input[['InputAmount']] = 'NA'
input[['Order Id']] = '4775'
input[['RawDataDir']] = 'p2886/NovaSeq_20180914_NOV19_o4775/B9P9_NPCHP11_NDd5'
input[['Read1']] = 'p2886/NovaSeq_20180914_NOV19_o4775/B9P9_NPCHP11_NDd5/B9P9_NPCHP11_NDd5_S5_L001_R2_001.fastq.gz'
input[['Read Count']] = '194401199'
debug(ezMethodCellRanger)
EzAppCellRanger$new()$run(input=input, output=output, param=param)

## p2284
setwd("/scratch/gtan/dev/cellranger-2884")
library(ezRun)
param = list()
param[['cores']] = '8'
param[['ram']] = '32'
param[['scratch']] = '100'
param[['node']] = ''
param[['process_mode']] = 'SAMPLE'
param[['name']] = 'CellRangerCount_NS223_o480'
param[['refBuild']] = 'Homo_sapiens/Ensembl/GRCh38.p10/Annotation/Release_91-2018-02-26'
param[['refFeatureFile']] = 'genes.gtf'
param[['specialOptions']] = ''
param[['mail']] = 'ge.tan@fgcz.ethz.ch'
param[['dataRoot']] = '/srv/gstore/projects'
param[['resultDir']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49'
output = list()
output[['Name']] = 'Recurrent_GBM_04092018'
output[['ResultDir [File]']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49/Recurrent_GBM_04092018'
output[['Report [Link]']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49/Recurrent_GBM_04092018/outs/web_summary.html'
output[['BAM [File]']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49/Recurrent_GBM_04092018/outs/possorted_genome_bam.bam'
output[['BAI [File]']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49/Recurrent_GBM_04092018/outs/possorted_genome_bam.bam.bai'
output[['Species']] = 'Homo sapiens (human)'
output[['reference']] = '10X_Ref_Human_GRCh38.p10_20180305_Release_91'
output[['CountMatrix [File]']] = 'p2284/CellRangerCount_29702_2018-09-20--10-32-49/Recurrent_GBM_04092018/outs/filtered_gene_bc_matrices/10X_Ref_Human_GRCh38.p10_20180305_Release_91/matrix.mtx'
output[['Sample Id [B-Fabric]']] = 'bfs_184413'
output[['Order Id [B-Fabric]']] = '4801'
input = list()
input[['Name']] = 'Recurrent_GBM_04092018'
input[['Condition']] = ''
input[['Species']] = 'Homo sapiens (human)'
input[['FragmentSize']] = '0'
input[['SampleConc']] = '1'
input[['Tube']] = 'p2284_4801/2'
input[['Sample Id']] = 'bfs_184413'
input[['PlatePosition']] = 'D1'
input[['LibConc_100_800bp']] = '30.8'
input[['LibConc_qPCR']] = '0'
input[['RIN']] = '0'
input[['Adapter1']] = 'NA'
input[['Adapter2']] = 'NA'
input[['strandMode']] = 'NA'
input[['LibraryPrepKit']] = '10X Genomics Single Cell'
input[['EnrichmentMethod']] = 'None'
input[['InputAmount']] = 'NA'
input[['Order Id']] = '4801'
input[['RawDataDir']] = 'p2284/NovaSeq_20181010_NOV28_o4801/Recurrent_GBM_04092018,p2284/NextSeq500_20180919_NS223_o4801/Recurrent_GBM_04092018'
input[['Read1']] = 'p2284/NovaSeq_20181010_NOV28_o4801/Recurrent_GBM_04092018/Recurrent_GBM_04092018_S1_L001_R2_001.fastq.gz,p2284/NextSeq500_20180919_NS223_o4801/Recurrent_GBM_04092018/Recurrent_GBM_04092018_S4_L001_R2_001.fastq.gz'
debug(ezMethodCellRanger)
EzAppCellRanger$new()$run(input=input, output=output, param=param)

# p2529 with control sequences
setwd("/scratch/gtan/dev/cellranger-p2529")
library(ezRun)
param = list()
param[['cores']] = '8'
param[['ram']] = '40'
param[['scratch']] = '200'
param[['node']] = ''
param[['process_mode']] = 'SAMPLE'
param[['samples']] = 'FAPS30_all_E1'
param[['name']] = 'CellRangerCount'
param[['refBuild']] = 'Mus_musculus/Ensembl/GRCm38.p5/Annotation/Release_91-2018-02-26'
param[['refFeatureFile']] = 'genes.gtf'
param[['featureLevel']] = 'gene'
param[['controlSeqs']] = 'GFP'
param[['specialOptions']] = ''
param[['mail']] = 'ge.tan@fgcz.ethz.ch'
param[['dataRoot']] = '/srv/gstore/projects'
param[['resultDir']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55'
output = list()
output[['Name']] = 'FAPS30_all_E1'
output[['ResultDir [File]']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55/FAPS30_all_E1'
output[['Report [Link]']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55/FAPS30_all_E1/outs/web_summary.html'
output[['BAM [Link]']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55/FAPS30_all_E1/outs/possorted_genome_bam.bam'
output[['BAI [Link]']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55/FAPS30_all_E1/outs/possorted_genome_bam.bam.bai'
output[['Species']] = 'Mus musculus (house mouse)'
output[['refBuild']] = 'Mus_musculus/Ensembl/GRCm38.p5/Annotation/Release_91-2018-02-26'
output[['refFeatureFile']] = 'genes.gtf'
output[['featureLevel']] = 'gene'
output[['CountMatrix [Link]']] = 'p2529/CellRangerCount_25013_2019-01-07--12-51-55/FAPS30_all_E1/outs/filtered_feature_bc_matrix'
output[['Sample Id [B-Fabric]']] = 'bfs_170597'
output[['Order Id [B-Fabric]']] = '4021'
input = list()
input[['Name']] = 'FAPS30_all_E1'
input[['Condition']] = ''
input[['Species']] = 'Mus musculus (house mouse)'
input[['FragmentSize']] = '0'
input[['SampleConc']] = '1'
input[['Tube']] = 'p2529_4021/1'
input[['Sample Id']] = 'bfs_170597'
input[['PlatePosition']] = ''
input[['LibConc_100_800bp']] = '70'
input[['LibConc_qPCR']] = '0'
input[['Adapter1']] = 'NA'
input[['Adapter2']] = 'NA'
input[['strandMode']] = 'NA'
input[['LibraryPrepKit']] = '10X Genomics Single Cell'
input[['EnrichmentMethod']] = 'None'
input[['InputAmount']] = 'NA'
input[['Order Id']] = '4021'
input[['RawDataDir']] = 'p2529/HiSeq2500_20180227_RR121_o4021/FAPS30_all_E1'
debug(getCellRangerReference)
debug(ezMethodCellRanger)
EzAppCellRanger$new()$run(input=input, output=output, param=param)

## p3360 tar
setwd("/scratch/gtan/debug/quickDebug")
library(ezRun)
param = list()
param[['cores']] = '32'
param[['ram']] = '40'
param[['scratch']] = '200'
param[['node']] = 'fgcz-c-048,fgcz-h-004,fgcz-h-006,fgcz-h-007,fgcz-h-008,fgcz-h-009,fgcz-h-010,fgcz-h-011,fgcz-h-012,fgcz-h-013,fgcz-h-014,fgcz-h-015,fgcz-h-016,fgcz-h-017,fgcz-h-018,fgcz-h-019,fgcz-h-020,fgcz-h-021,fgcz-h-022,fgcz-h-023,fgcz-h-024,fgcz-h-025'
param[['process_mode']] = 'SAMPLE'
param[['samples']] = 'RT1,Abx5,C3,CAX6,RT5,Abx6,C7,CAX5'
param[['name']] = 'CellRangerCount'
param[['refBuild']] = 'Mus_musculus/GENCODE/GRCm38.p6/Annotation/Release_M23-2019-11-05'
param[['refFeatureFile']] = 'genes.gtf'
param[['featureLevel']] = 'gene'
param[['TenXLibrary']] = 'GEX'
param[['scMode']] = 'SN'
param[['chemistry']] = 'auto'
param[['transcriptTypes']] = 'protein_coding,rRNA,tRNA,Mt_rRNA,Mt_tRNA'
param[['controlSeqs']] = ''
param[['cmdOptions']] = ''
param[['specialOptions']] = ''
param[['mail']] = 'ge.tan@fgcz.ethz.ch'
param[['dataRoot']] = '/srv/gstore/projects'
param[['resultDir']] = 'p3360/CellRangerCount_44147_2020-02-18--17-16-35'
param[['isLastJob']] = TRUE
output = list()
output[['Name']] = 'CAX5'
output[['Species']] = 'Mus musculus (house mouse)'
output[['refBuild']] = 'Mus_musculus/GENCODE/GRCm38.p6/Annotation/Release_M23-2019-11-05'
output[['refFeatureFile']] = 'genes.gtf'
output[['featureLevel']] = 'gene'
output[['transcriptTypes']] = 'protein_coding,rRNA,tRNA,Mt_rRNA,Mt_tRNA'
output[['ResultDir [File]']] = 'p3360/CellRangerCount_44147_2020-02-18--17-16-35/CAX5'
output[['Report [Link]']] = 'p3360/CellRangerCount_44147_2020-02-18--17-16-35/CAX5/web_summary.html'
output[['CountMatrix [Link]']] = 'p3360/CellRangerCount_44147_2020-02-18--17-16-35/CAX5/filtered_feature_bc_matrix'
output[['Sample Id [B-Fabric]']] = 'bfs_213909'
output[['Order Id [B-Fabric]']] = '6714'
input = list()
input[['Name']] = 'CAX5'
input[['Condition']] = ''
input[['Species']] = 'Mus musculus (house mouse)'
input[['FragmentSize']] = '0'
input[['SampleConc']] = '0'
input[['Tube']] = 'p3360_6714/8'
input[['Sample Id']] = 'bfs_213909'
input[['PlatePosition']] = ''
input[['LibConc_100_800bp']] = '0'
input[['LibConc_qPCR']] = '0'
input[['RIN']] = '0'
input[['Adapter1']] = 'NA'
input[['Adapter2']] = 'NA'
input[['strandMode']] = 'NA'
input[['LibraryPrepKit']] = '10X Genomics Single Cell, V3.1'
input[['EnrichmentMethod']] = 'None'
input[['InputAmount']] = 'NA'
input[['Order Id']] = '6714'
input[['PlateName']] = 'p3360_o6714'
input[['RawDataDir']] = 'p3360/NovaSeq_20200217_NOV321_o6714_DataDelivery/CAX5.tar'
input[['Read1']] = 'p3360/NovaSeq_20200217_NOV321_o6714_DataDelivery/CAX5/CAX5_S8_L001_R2_001.fastq.gz'
input[['Read Count']] = '238527372'
debug(ezMethodCellRanger)
EzAppCellRanger$new()$run(input=input, output=output, param=param)
