---
title: "`r paste('Metagenome Atlas Summary')`"
author: "Functional Genomics Center Zurich"
output: 
    html_document:
    self_contained: false
lib_dir: rmarkdownLib
includes:
    in_header: fgcz_header.html
css: fgcz.css
editor_options: 
    chunk_output_type: console
---

    
    ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, out.width = "49%")
## This report requires the output of Metagenome Atlas app
debug <- FALSE
```

# {.tabset}

## Quality control
```{r QC, echo=FALSE}
htmltools::includeHTML(QCHTML)
```

<!-- ## Assembly report -->
<!-- ```{r assembly, echo=FALSE} -->
<!-- htmltools::includeHTML(ASSEMBLYHTML) -->
<!-- ``` -->

<!-- ## Binning report -->
<!-- ```{r bin, echo=FALSE} -->
<!-- htmltools::includeHTML(BINNINGHTML) -->
<!-- ``` -->

<!-- ## Bins to MAGs and more -->

<!-- ```{r, echo=FALSE} -->
<!-- DT::datatable(genomic_bins_table) -->
<!-- ``` -->

<!-- ```{r, echo=FALSE} -->
<!-- DT::datatable(taxonomy_table) -->
<!-- ``` -->

<!-- ```{r, echo=FALSE} -->
<!-- myTree <- ape::read.tree(text=newick) -->
<!-- plot(myTree,no.margin=TRUE,edge.width=2,cex=0.9) -->
<!-- ``` -->

