---
title: "`r paste('Shotgun metagenomics analysis report')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, out.width = "33%")
## This report requires summary files form mothur
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepare data, include=FALSE}
##Prodigal
### summary  plots : score, gc_cont abd confidence
summaryScorePlot <- ggplot(prodigalSummaryDF,aes(x=score)) + geom_histogram(binwidth=10) +  facet_grid(rows = vars(start_type), cols = vars(partial)) + labs(title="Summary of the gene prediction scores")

subsetDataToPartial00 <- prodigalSummaryDF[prodigalSummaryDF$partial =="00" 
                                           & (prodigalSummaryDF$start_type == "ATG"|
                                                prodigalSummaryDF$start_type == "GTG"),]
summaryConfPlot <- ggplot(subsetDataToPartial00,aes(x=conf)) + geom_histogram(binwidth=5) +  facet_grid(rows = vars(start_type), cols = vars(partial),labeller = label_both)+ labs(title="Summary of the gene prediction confidence")

summaryGcContPlot <- ggplot(subsetDataToPartial00,aes(x=gc_cont)) + geom_histogram(binwidth=0.001) +  facet_grid(rows = vars(start_type), cols = vars(partial),labeller = label_both) + labs(title="Summary of the GC-content")

### summary rbs_spacer hist
summaryRBSSpacePlot <- ggplot(subsetDataToPartial00,aes(x=rbs_spacer)) + geom_bar() +  facet_grid(rows = vars(start_type), cols = vars(partial), labeller = label_both) +theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  labs(title="RBS-spacer distribution")
### summary rbs_motif hist
summaryRBSMotifPlot <- ggplot(subsetDataToPartial00,aes(x=rbs_motif)) + geom_bar() +  facet_grid(rows = vars(start_type), cols = vars(partial), labeller = label_both) + theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(title="RBS-motif distribution")


## Inteproscan
### summary  plots : matching score 
summaryMatchScorePlot <- ggplot(IPSGffSummaryDF,aes(x=-log(score))) + geom_histogram(binwidth = 5) +
  labs(title="Summary of protein match score")

### topNcateg plots: tables
DFforSummProtFamilies <- extractTopN(IPSGffSummaryDF,"description",numberOfTopNCategories)
DFforSummProtGO <- extractTopN(IPSGffSummaryDF,"GOterm",numberOfTopNCategories)
## Add GO decsription
GOdesc <- sapply(DFforSummProtGO$GOterm,function(x)Term(GOTERM)[names(Term(GOTERM))%in%x], USE.NAMES = F)
DFforSummProtGO$GOterm <- paste(DFforSummProtGO$GOterm,GOdesc)
## expand palette colour to numberOfTopNCategories
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
expandedPalette <- getPalette(numberOfTopNCategories)
## generate plots
summaryFamilyPlot <- ggplot(DFforSummProtFamilies,aes(x=reorder(description, -abundance),y=abundance, fill = description)) 
summaryFamilyPlot <- summaryFamilyPlot + geom_bar(stat = "Identity")+theme(axis.text.x = element_blank(), axis.title.x = element_blank(), legend.position = "bottom",legend.text = element_text(size =7)) + scale_color_manual(expandedPalette)
summaryFamilyPlot <- summaryFamilyPlot + guides(fill=guide_legend(ncol=2, byrow=F,title.position = "top"))  +
  labs(title="Most represented  protein families")


summaryGOPlot <- ggplot(DFforSummProtGO,aes(x=reorder(GOterm, -abundance),y=abundance, fill = GOterm)) 
summaryGOPlot <- summaryGOPlot + geom_bar(stat = "Identity")+theme(axis.text.x = element_blank(), axis.title.x = element_blank(),legend.position = "bottom",legend.text = element_text(size =7)) + scale_color_manual(expandedPalette)
summaryGOPlot <- summaryGOPlot +  guides(fill=guide_legend(ncol=2, byrow=F,title.position = "top"))  +
  labs(title="Most represented GO terms")
```


Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Data pre-processing {.tabset}

### Prodigal gene prediction
#### Summary score plot
```{r summaryScorePlot, echo=FALSE}
par(mfrow=c(1,3), las=1)
plot(summaryScorePlot)
plot(summaryConfPlot)
plot(summaryGcContPlot)
plot(summaryRBSSpacePlot)
plot(summaryRBSMotifPlot)
```

### Interproscan annotation
#### Summary of matched proteins score
```{r summaryMatchScorePlot, echo=FALSE,out.width="33%"}
par(mfrow=c(1,2), las=1)
plot(summaryFamilyPlot)
plot(summaryGOPlot)
plot(summaryMatchScorePlot)
```