---
title: "`r paste('Mothur metagenomics workflow')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## This report requires summary files form mothur

require(knitr)
require(kableExtra)
require(SummarizedExperiment)
require(webshot)
require(htmlwidgets)
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepare data, include=FALSE}
### rawDataSummary
rawSummaryTableIllumina <- createSummaryTable(rawIllumina)
rawSummaryTableIllumina <- rawSummaryTableIllumina$rawDataSummaryTable

deduppedSummary <- createSummaryTable(deduppedIllumina)
deduppedSummaryTable <- deduppedSummary$rawDataSummaryTable
titleDeduppedSummaryTable<- deduppedSummary$rawDataSummaryTableTitle

lengthAndHomopFilteredSummary <- createSummaryTable(lengthAndHomopFilteredIllumina)
lengthAndHomopFilteredSummaryTable <- lengthAndHomopFilteredSummary$rawDataSummaryTable
titleLengthAndHomopFilteredSummaryTable <- lengthAndHomopFilteredSummary$rawDataSummaryTableTitle

mappedFilteredSummary <- createSummaryTable(mappedFilteredIllumina)
mappedFilteredSummaryTable <- mappedFilteredSummary$rawDataSummaryTable
titleMappedFilteredSummaryTable <- mappedFilteredSummary$rawDataSummaryTableTitle

preClusteredAndChimeraFilteredSummary <- createSummaryTable(preClusteredAndChimeraFilteredIllumina)
preClusteredAndChimeraFilteredSummaryTable <- preClusteredAndChimeraFilteredSummary$rawDataSummaryTable
titlePreClusteredAndChimeraFilteredSummaryTable <- preClusteredAndChimeraFilteredSummary$rawDataSummaryTableTitle

chimeraIlluminaPlot <- chimeraSummaryPlot(chimeraIllumina)
```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Data pre-processing {.tabset}

### Summary of the raw reads
```{r sumIll, echo=FALSE}
kable(rawSummaryTableIllumina, row.names=TRUE, 
      col.names=colnames(rawSummaryTableIllumina), format="html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads filtered by duplication
```{r filtIll, echo=FALSE}
kable(deduppedSummaryTable, row.names=TRUE, 
      col.names=colnames(deduppedSummaryTable), format="html",caption =titleDeduppedSummaryTable) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads filtered by length  and Homopolymers
```{r mapIll, echo=FALSE}
kable(lengthAndHomopFilteredSummaryTable, row.names=TRUE, 
      col.names=colnames(lengthAndHomopFilteredSummaryTable), format="html",caption =titleLengthAndHomopFilteredSummaryTable) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads filtered by mapping to the Silva database
```{r mapIll, echo=FALSE}
kable(mappedFilteredSummaryTable, row.names=TRUE, 
      col.names=colnames(mappedFilteredSummaryTable), format="html",caption =titleMappedFilteredSummaryTable) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads preclusterd and filtered for chimeras
```{r precIll, echo=FALSE}
kable(preClusteredAndChimeraFilteredSummaryTable, row.names=TRUE, 
      col.names=colnames(preClusteredAndChimeraFilteredSummaryTable), format="html", caption = titlePreClusteredAndChimeraFilteredSummaryTable) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
plot(chimeraIlluminaPlot)
```


### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
