---
title: "`r paste('Mothur metagenomics workflow')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, fig.width=10, fig.height=7)
## This report requires summary files form mothur

require(knitr)
require(kableExtra)
require(SummarizedExperiment)
require(webshot)
require(htmlwidgets)
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepare data AB, include=FALSE}
### create plots: 1. abundance
mergedAbPlotIndPhylum <- plot_bar(physeqMerged, fill="Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")
mergedAbPlotGroupClass <- plot_bar(physeqMerged, "Group", "Abundance", fill="Class") + geom_bar(aes(color=Class, fill=Class), stat="identity", position="stack")
```

```{r prepare data COMP, include=FALSE}
### Comparison with mock community
mockPacbioPlot <- phyloSeqCommunityComp(physeqPacBioNoTree)
mockIlluminaPlot <- phyloSeqCommunityComp(physeqIllNoTree)
```

```{r prepare data ORD, include=FALSE}
### create plots: 2. ordination
mergedOrd <- ordinate(physeqMerged, "NMDS", "bray")
plotOrdTaxaMerged = plot_ordination(physeqMerged, mergedOrd, type="taxa", color="Phylum") + facet_wrap(~Phylum, 3) + ggtitle("Taxa ordination, all samples")
plotOrdSamplesMerged = plot_ordination(physeqMerged, mergedOrd, type="samples", color="Group", shape = "Technology") 
plotOrdSamplesMerged = plotOrdSamplesMerged + geom_polygon(aes(fill=Group)) + geom_point(size=3) + ggtitle("Sample ordination, all samples")

#illOrd <- ordinate(physeqIllNoTree, "NMDS", "bray")
#plotOrdTaxaIll = plot_ordination(physeqIllNoTree, illOrd, type="taxa", color="Phylum") + facet_wrap(~Phylum, 3) + #ggtitle("Taxa ordination, only Illumina samples")
#plotOrdSamplesIll = plot_ordination(physeqIllNoTree, illOrd, type="samples", color="Group") 
#plotOrdSamplesIll = plotOrdSamplesIll + geom_polygon(aes(fill=Group)) + geom_point(size=5) + ggtitle("Sample ordination, only #Illumina samples")
```

```{r prepare data RICH, include=FALSE}
### create plots: 3. richness 
plotRichMerged <- plot_richness(physeqMerged, x="Technology", color = "Group", measures=c("Chao1", "Shannon"))
```

```{r prepare data DIST, include=FALSE}
### create plots: distribution deepest taxa
plotDeepDistPB  <- phyloSeqDivPlotAndPercUnclassified(basename(input$getFullPaths("Taxonomy_pacbio")),"Genus")
plotDeepDistIll <- phyloSeqDivPlotAndPercUnclassified(basename(input$getFullPaths("Taxonomy_Illumina")),"Genus")
```

```{r prepare data TREE, include=FALSE}
### create plots: 4. tree
plotTreeMerged <- plot_tree(physeqMergedPruned , ladderize="left", color="Group", shape="Technology")
onlyBacteroidetes <- subset_taxa(physeqMergedPruned, Phylum=="Bacteroidetes")
plotTreeBacteroidetes <-  plot_tree(onlyBacteroidetes , ladderize="left", color="Group", shape="Technology", size = "abundance", label.tips="Genus") + ggtitle("Bacteroidetes-only tree")
```

```{r prepare data HEAT, include=FALSE}
### create plots:5. heatmap
plotHeatmapMerged <- plot_heatmap(physeqMergedPruned, taxa.label="Genus")
```

```{r prepare data deseq, include=FALSE}
### 6: compare groups only Illumina for now
deseqResults <- phyloSeqToDeseq2_tableAndPlots(physeqIllNoTree)
tableToPlot <- deseqResults$table
tableToPlot <- na.omit(tableToPlot[tableToPlot$padj < 0.05,])
tableToPlot <- tableToPlot[order(tableToPlot$padj),]

```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Data anlysis and visualization with phyloseq {.tabset}

### Abundance barplots
#### Community composition by sample: phyla
```{r abundPBSamplePhylum, echo=FALSE}
plot(mergedAbPlotIndPhylum)
```

#### Community composition by group: class
```{r abundPBGroupClass, echo=FALSE}
plot(mergedAbPlotGroupClass)
```

### Comparison with the mock community
#### PacBio
```{r mockPB, echo=FALSE}
plot(mockPacbioPlot)
```

####Illumina
```{r mockIll, echo=FALSE}
plot(mockIlluminaPlot)
```

### Ordination plots
#### Taxa level
```{r ordPBtaxa, echo=FALSE}
plot(plotOrdTaxaMerged)
plot(plotOrdTaxaIll)
```

#### Sample level
```{r ordPB, echo=FALSE}
plot(plotOrdSamplesMerged)
plot(plotOrdSamplesIll)

```

### Richness plots
#### Community alpha-diversity
```{r richPB, echo=FALSE}
plot(plotRichMerged)
```

### Distribution of the deepest taxa
#### PacBio
```{r distPB, echo=FALSE}
plot(plotDeepDistPB)
```

#### Illumina
```{r distIll, echo=FALSE}
plot(plotDeepDistIll)
```

### Tree plots
#### Full tree
```{r treePB, echo=FALSE}
plot(plotTreeMerged)
```

#### Subsetted tree
```{r radialTreePB, echo=FALSE}
plot(plotTreeBacteroidetes)
```

### Heatmaps
#### Abundances
```{r heatPB, echo=FALSE}
plot(plotHeatmapMerged)
```

### Two-goups comparison (Illumina real data)
#### Volcano plot
```{r volcano, echo=FALSE}
plot(deseqResults$vPlot)
```

#### Log-fold changes 
```{r logfold, echo=FALSE}
plot(deseqResults$logPlot)
```

#### Summary table (Significant OTUs)
```{r table, echo=FALSE}
kable(tableToPlot, row.names=FALSE, 
      col.names=colnames(tableToPlot), format="html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

#### Differentially adundant Genera 
```{r DiffAbundGenPie, echo=FALSE}
plot(deseqResults$pieChart)
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
