---
title: "`r paste('Mothur metagenomics workflow')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## This report requires summary files form mothur

require(knitr)
require(kableExtra)
require(SummarizedExperiment)
require(webshot)
require(htmlwidgets)
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepare data, include=FALSE}
### rawDataSummary
rawSummaryTableIllumina <- createSummaryTable(rawIllumina)
rawSummaryTableIllumina <- rawSummaryTableIllumina$rawDataSummaryTable

lengthDeduppedSummaryTableIllumina <- createSummaryTable(lengthDeduppedIllumina)
lengthDeduppedSummaryTableIllumina <- lengthDeduppedSummaryTableIllumina$rawDataSummaryTable
titleLengthDeduppedSummaryTableIllumina <- createSummaryTable(lengthDeduppedIllumina)
titleLengthDeduppedSummaryTableIllumina <- titleLengthDeduppedSummaryTableIllumina$rawDataSummaryTableTitle

mappedAndHomopFilteredSummaryTableIllumina <- createSummaryTable(mappedAndHomopFilteredIllumina)
mappedAndHomopFilteredSummaryTableIllumina <- mappedAndHomopFilteredSummaryTableIllumina$rawDataSummaryTable
titleMappedAndHomopFilteredSummaryTableIllumina <- createSummaryTable(mappedAndHomopFilteredIllumina)
titleMappedAndHomopFilteredSummaryTableIllumina <- titleMappedAndHomopFilteredSummaryTableIllumina$rawDataSummaryTableTitle

preClusteredAndChimeraFilteredSummaryTableIllumina <- createSummaryTable(preClusteredAndChimeraFilteredIllumina)
preClusteredAndChimeraFilteredSummaryTableIllumina <- preClusteredAndChimeraFilteredSummaryTableIllumina$rawDataSummaryTable
titlePreClusteredAndChimeraFilteredSummaryTableIllumina <- createSummaryTable(preClusteredAndChimeraFilteredIllumina)
titlePreClusteredAndChimeraFilteredSummaryTableIllumina <- titlePreClusteredAndChimeraFilteredSummaryTableIllumina$rawDataSummaryTableTitle

chimeraIlluminaPlot <- chimeraSummaryPlot(chimeraIllumina)
```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Data pre-processing {.tabset}

### Summary of the raw reads
```{r sumIll, echo=FALSE}
kable(rawSummaryTableIllumina, row.names=TRUE, 
      col.names=colnames(rawSummaryTableIllumina), format="html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads filtered by length and duplication
```{r filtIll, echo=FALSE}
kable(lengthDeduppedSummaryTableIllumina, row.names=TRUE, 
      col.names=colnames(lengthDeduppedSummaryTableIllumina), format="html",caption =titleLengthDeduppedSummaryTableIllumina) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads filtered by mapping coordinates and Homopolymers

```{r mapIll, echo=FALSE}
kable(mappedAndHomopFilteredSummaryTableIllumina, row.names=TRUE, 
      col.names=colnames(preClusteredAndChimeraFilteredSummaryTablePacbio), format="html",caption =titleMappedAndHomopFilteredSummaryTableIllumina) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

### Reads preclusterd and filtered for chimeras
```{r precIll, echo=FALSE}
kable(preClusteredAndChimeraFilteredSummaryTableIllumina, row.names=TRUE, 
      col.names=colnames(preClusteredAndChimeraFilteredSummaryTableIllumina), format="html", caption = titlePreClusteredAndChimeraFilteredSummaryTableIllumina) %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
plot(chimeraIlluminaPlot)
```


### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
