---
title: "`r paste('Analysis of 16S data: phyloseq report')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, out.width = "49%")
## This report requires summary files form mothur
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepAbund, include=FALSE}
### create plots: 1. abundance
abundPlotList <- list()
taxRanksVec <- attr(physeqFullObject@tax_table@.Data,"dimnames")[[2]]
taxRanks <- setNames(as.list(taxRanksVec),taxRanksVec)
abundPlotList <- lapply(taxRanks,abundPlot,physeqFullObject=physeqFullObject,xAes="Sample",y="relSampleFract")

if (isGroupThere) {
  abundPlotListGroup <- lapply(taxRanks,abundPlot,physeqFullObject=physeqFullObject,xAes="group",y="relGroupFract")
}
```

```{r prepOrdTaxa, include=FALSE, eval=isGroupThere}
### create plots: 2. ordination by taxa
ordPlotSampleList <- lapply(taxRanks,ordPlot,physeqFullObject=physeqFullObject, type="samples")
```

```{r prepOrdSamples, include=FALSE, eval=isGroupThere}
### create plots: 2. ordination by samlpe
ordPlotSampleList <- lapply(taxRanks,ordPlot,physeqFullObject=physeqFullObject, type="taxa")
```

```{r prepRich, include=FALSE}
### create plots: 3. richness 
plotRichnessBySample <- plot_richness(physeqFullObject, measures=c("Shannon"))
if (isGroupThere) {
plotRichnessByGroup <- groupModRichPlot(physeqFullObject, x="group", measures=c("Shannon"))
}
```


```{r prepHeat, include=FALSE, eval=isGroupThere}
### create plots:5. pheatmap
show_pHeatmap <- heatmapForPhylotseqPlotPheatmap(inputData,designMatrix)
```


```{r prepDeSeq, include=FALSE, eval=isGroupThere}
### 6: compare groups
deseqResults <- phyloSeqToDeseq2_tableAndPlots(physeqFullObject)
```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Phyloseq report {.tabset}
### Abundance barplots
#### If samples are grouped, the second plot  is more informative
```{r abund, echo=FALSE}
par(mfrow=c(1,2), las=1)
plot(abundPlot)
if (isGroupThere) {
plot(abundPlotGroup)
}
```

### Clustering (ordination) plots
```{r ordinTaxa, echo=FALSE,eval=isGroupThere}
par(mfrow=c(1,2), las=1)
plot(plotOrdTaxa)
plot(plotOrdSamples)
```

### Richness plots
```{r rich, echo=FALSE,out.width="60%"}
plot(plotRichness)
```

```{r tree, echo=FALSE, eval=FALSE}
plot(plotTree)
```

### Rarefaction plots
```{r raref, echo=FALSE}
par(mfrow=c(1,2), las=1)
for (RarPlot in rarefPlot){
plot(RarPlot)
}
```

### Heatmaps
```{r heat, echo=FALSE,eval=isGroupThere,out.width="100%"}
show_pHeatmap()
```

### Two-goups comparison
```{r volcano, echo=FALSE,eval=isGroupThere, out.width="33%"}
par(mfrow=c(1,3), las=1)
plot(deseqResults$vPlot)
plot(deseqResults$logPlot)
plot(deseqResults$pieChart)
```

#### Summary table (top 20 differentially expressed organisms)
```{r deSeqTable, echo=FALSE,eval=isGroupThere}
kable(deseqResults$table, escape=TRUE, row.names=FALSE, format="html") %>%
  kable_styling(bootstrap_options = c("striped", "bordered"),
                full_width = T) %>%
  scroll_box(width = "100%", height = "600px")
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
