---
title: "`r paste('Analysis of 16S data: phyloseq report')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, out.width = "49%")
## This report requires summary files form mothur
debug <- FALSE
setEnvironments("phantomjs")
```

```{r prepAbund, include=FALSE}
### create plots: 1. abundance
abundPlotSampleList <- list()
abundPlotGroupList <- list()
abundPlotSampleListUnfilt <- list()
abundPlotGroupListUnfilt <- list()
taxRanksVec <- attr(physeqFullObject@tax_table@.Data,"dimnames")[[2]]
taxRanks <- setNames(as.list(taxRanksVec),taxRanksVec)
for (rank  in taxRanks){
  tempAbundPlotUnfilt <-  abundPlot(rank,physeqObjectNoTreeUnfilt,xAes="Sample",y="relSampleFract")
  tempAbundPlotFilt <- abundPlot(rank,physeqFullObject,xAes="Sample",y="relSampleFract")
  if (tempAbundPlotFilt$stop==TRUE) {
    next
  } else {
abundPlotSampleList[[rank]] <- tempAbundPlotFilt$abPlot 
  }
    if (tempAbundPlotUnfilt$stop==TRUE) {
    next
  } else {
abundPlotSampleListUnfilt[[rank]] <- tempAbundPlotUnfilt$abPlot
  }
}

if (isGroupThere) {
for (rank  in taxRanks){
  tempAbundPlotUnfilt <-  abundPlot(rank,physeqObjectNoTreeUnfilt,xAes="Sample",y="relGroupFract")
  tempAbundPlotFilt <- abundPlot(rank,physeqFullObject,xAes="Sample",y="relGroupFract")
  
  if (tempAbundPlotFilt$stop==TRUE) {
    next
  } else {
abundPlotGroupList[[rank]] <- tempAbundPlotFilt$abPlot
  }
    if (tempAbundPlotUnfilt$stop==TRUE) {
    next
  } else {
abundPlotGroupListUnfilt[[rank]] <- tempAbundPlotUnfilt$abPlot
  }
}
}
isAbundPlotSample <- !length(abundPlotSampleList) == FALSE  
isAbundPlotGroup <-!length(abundPlotGroupList) == FALSE
isAbundPlotSampleUnfilt <- !length(abundPlotSampleListUnfilt) == FALSE  
isAbundPlotGroupUnfilt <-!length(abundPlotGroupListUnfilt) == FALSE
```

```{r prepOrdTaxa, include=FALSE}
### create plots: 2. ordination by taxa
ordPlotTaxa <- ordPlot(rank,physeqFullObject,"taxa")
```

```{r prepOrdSamples, include=FALSE, eval=isGroupThere}
### create plots: 2. ordination by samlpe
ordPlotSample<- ordPlot(rank,physeqFullObject,"samples",areThereMultVar)
```

```{r prepRich, include=FALSE}
### create plots: 3. richness. Here we use only the unfiltered dataset.
plotRichnessBySample <- plot_richness(physeqObjectNoTreeUnfilt, measures=c("Shannon"))
if (isGroupThere) {
plotRichnessByGroup <- groupModRichPlot(physeqObjectNoTreeUnfilt, x="group", measures=c("Shannon"))
}
```


```{r prepHeat, include=FALSE, eval=isGroupThere}
### create plots:5. pheatmap
show_pHeatmap <- heatmapForPhylotseqPlotPheatmap(physeqFullObject,areThereMultVar)
```


```{r prepDeSeq, include=FALSE, eval=isGroupThere}
### 6: compare groups
deseqResults <- phyloSeqToDeseq2_tableAndPlots(physeqFullObject,rank)
```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Phyloseq report {.tabset}
### Abundance barplots
#### If samples are grouped, the second plot  is more informative
```{r abund, echo=FALSE}
par(mfrow=c(2,2), las=1)
if (isAbundPlotSample) {
plot(abundPlotList[[rank]])
}else{
  message <- paste("After filtering, no taxa remaining wih a", rank,"annotaton. Nothing to show.")
  printf(message)
}
if (isAbundPlotSampleUnfilt) {
plot(abundPlotSampleListUnfilt[[rank]])
}else{
  message <- paste("No taxa annotated at the", rank,"level Nothing to show.")
  printf(message)
}


if (isGroupThere) {
  if (isAbundPlotGroup) {
plot(abundGroupList[[rank]])
  }else{
    message <- paste("After filtering, no taxa remaining wih a", rank,"annotaton. Nothing to show.")
  printf(message)  
  }
  if (isAbundPlotGroupUnfilt) {
plot(abundPlotGroupListUnfilt[[rank]])
}else{
  message <- paste("No taxa annotated at the", rank,"level Nothing to show.")
  printf(message)
}

}
```

### Clustering (ordination) plots
```{r ordinTaxa, echo=FALSE, }
par(mfrow=c(1,2), las=1)
plot(ordPlotTaxa)
if (isGroupThere) {
plot(ordPlotSample)
}
```

### Richness plots
```{r rich, echo=FALSE}
par(mfrow=c(1,2), las=1)
plot(plotRichnessBySample)
if (isGroupThere) {
  plot(plotRichnessByGroup)
}
```

### Rarefaction plots
```{r raref, echo=FALSE, eval=FALSE}
par(mfrow=c(1,2), las=1)
for (RarPlot in rarefPlot){
plot(RarPlot)
}
```

### Heatmaps
```{r heat, echo=FALSE,eval=isGroupThere,out.width="100%"}
show_pHeatmap()
```

### Two-goups comparison
```{r volcano, echo=FALSE,eval=isGroupThere, out.width="33%"}
par(mfrow=c(1,3), las=1)
plot(deseqResults$vPlot)
plot(deseqResults$logPlot)
plot(deseqResults$pieChart)
```

#### Summary table (top 20 differentially expressed organisms)
```{r deSeqTable, echo=FALSE,eval=isGroupThere}
kable(deseqResults$table, escape=TRUE, row.names=FALSE, format="html") %>%
  kable_styling(bootstrap_options = c("striped", "bordered"),
                full_width = T) %>%
  scroll_box(width = "100%", height = "600px")
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
