---
title: "`r if (exists('reportTitle')) reportTitle else 'SUSHI Report'`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: true
    includes:
     in_header: !expr system.file("templates/fgcz_header.html", package="ezRun", lib.loc=.libPaths())
    css: !expr system.file("templates/fgcz.css", package="ezRun", lib.loc=.libPaths())
editor_options: 
  chunk_output_type: console
---
  
  Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, include=FALSE}
library(ggridges)
library(cowplot)
library(kableExtra)
library(pheatmap)
library(tidyverse)
library(ezRun)
library(HDF5Array)
library(scater)
library(scanalysis)

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, knitr.table.format = "html")
```

```{r}
input = readRDS("input.rds")
param = readRDS("param.rds")
URLs <- input$getColumn("Static Report")
```


```{r read sce, eval=param$samples == "SCReportMultipleSamples"}
filePath <- file.path("/srv/gstore/projects", sub("https://fgcz-(gstore|sushi).uzh.ch/projects", "",dirname(URLs)), "sce_h5")
filePath_course <- file.path(paste0("/srv/GT/analysis/course_sushi/public/projects/", input$getColumn("Report"), "/sce_h5"))

  if(file.exists(filePath)) {
    filePath <- filePath
 } else {
    filePath <- filePath_course
 }  
sce <- loadHDF5SummarizedExperiment(filePath)
```

```{r read seurat, eval=param$samples == "SpatialSeuratSlides"}
filePath <- file.path("/srv/gstore/projects", sub("https://fgcz-(gstore|sushi).uzh.ch/projects", "",dirname(URLs)), "scData.rds")
scData <- readRDS(filePath)
sce <- seurat_to_sce(scData)
```


```{r}
current_cells=read_tsv("current_cells.tsv")$cell
sce = sce[,current_cells]
sce$seurat_clusters = droplevels(sce$seurat_clusters)
var_heigth <- 1
```


```{r javascript, results='asis'}
jsFile = system.file("extdata/enrichr.js", package="ezRun", mustWork=TRUE)
invisible(file.copy(from=jsFile, to=basename(jsFile), overwrite=TRUE))
cat(paste0("<SCRIPT language=\"JavaScript\" SRC=\"", basename(jsFile), "\"></SCRIPT>"))
```

## Analysis results {.tabset}


```{r}
conservedMarkers <- tryCatch({read_tsv("conserved_markers.tsv")},error = function(e) NULL)
diffGenes <- tryCatch({read_tsv("differential_genes.tsv")},error = function(e) NULL)
```



```{r consMarkers table, eval=!is.null(conservedMarkers), results='asis'}
cat("### Conserved markers")
cat("\n")
cat("Identify cell type marker genes that are conserved across conditions. Differential gene expression tests are performed for each group and then, the p-values are combined using meta-analysis methods.")
cat("\n")
cat("#### Conserved cell type markers")
cat("\n")
ezInteractiveTableRmd(conservedMarkers)
```

```{r, eval=!is.null(conservedMarkers)}
top5 <- conservedMarkers %>% group_by(cluster)
top5 <- slice_max(top5, n = 5, order_by = avg_avg_fc)
var_heigth <- nrow(top5)*0.2
```


```{r consMarkers plots, eval=!is.null(conservedMarkers), results='asis', fig.width=8, fig.height=var_heigth}
cat("### Conserved markers plots\n")
cat("\n")
cat("Here, we use a heatmap and a dotplot to visualize simultaneously the top 5 conserved markers in each cluster (with highest average fold change across the two groups). Be aware that some genes may be in the top markers for different clusters.\n")
cat("\n")
plotHeatmap(sce, features=top5$gene, center=TRUE, zlim=c(-2, 2),
            colour_columns_by = "seurat_clusters", 
            order_columns_by = "seurat_clusters",
            cluster_rows=TRUE,
            fontsize_row=8)
plotDots(sce, features=top5$gene, group="seurat_clusters")
```

```{r diffGenes table, eval=!is.null(diffGenes), results='asis'}
cat("### Differential expressed genes")
cat("\n")

cat(paste0("After identifying common cell types across conditions, we can look for genes that change in different conditions for cells of the same type. The interactive table can help you to explore the more relevant genes (maximum 1000 genes with the biggest diff_pct). The full list of tested genes is in the Data availability section of this report. The comparison made was ", paste(c(param$refGroup,param$sampleGroup), collapse = " vs ")))

cat("\n")

diffGenes <- diffGenes[diffGenes$p_val_adj<0.05,]  #too many genes make the interactive table slow
cutoff <- 1000

if(nrow(diffGenes) > cutoff)
   diffGenes <- diffGenes[1:cutoff,]

cat("#### Differential expressed genes per cluster")
cat("\n")
ezInteractiveTableRmd(diffGenes)
```

```{r diffGenes heatmap, eval=!is.null(diffGenes), results='asis', fig.height=6, fig.width=12, fig.align='center'}
cat("### Differential genes plots\n")
cat("\n")
cat("The 50 dysregulated genes with the highest difference in expression between the two conditions are represented in the heatmap and the violin plots. The heatmap shows global differences of these genes across the conditions, while the violin plots show cluster specific changes across conditions.")
cat("\n")
most_diffGenes <- head(unique(diffGenes$gene), 50)
plotHeatmap(sce, order_columns_by = "Condition", features = most_diffGenes, fontsize=8)
```


```{r diffGenes violin plots, eval=!is.null(diffGenes), fig.height=30, fig.width=16}
plotExpression(sce, most_diffGenes, x="ident", colour_by="Condition", ncol=3)
```


### Data availability

##### Conserved markers 

[conservedMarkers](conserved_markers.tsv)

##### Differentially expressed genes

[diffGenes](differential_genes.tsv)

### Parameters
```{r report parameters, echo=FALSE}
param[c("DE.method", "DE.regress")]
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
