---
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: true
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
editor_options: 
  chunk_output_type: console
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, include=FALSE}
library(SummarizedExperiment)
library(ggplot2)
library(ezRun)
library(cowplot)
library(tibble)
library(dplyr)
library(readr)
library(kableExtra)
library(tidyr)
library(pheatmap)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, knitr.table.format = "html")


#to test
# param = readRDS("/srv/gstore/projects/p2979/SCMultipleSamplesOneGroup_40181_2019-10-09--17-06-26/SCReportMerging/param.rds")
# scData_objects = readRDS("/srv/gstore/projects/p2979/SCMultipleSamplesOneGroup_40181_2019-10-09--17-06-26/SCReportMerging/scData_ObjectList.rds")
# scData_noCorrected = scData_objects$scData_noCorrected
# scData = scData_objects$scData_corrected
# library("Seurat", lib="/home/daymegr/myRpackages")
# library(ggplot2)
# pvalue_allMarkers <- 0.05
```

```{r javascript, results='asis'}
jsFile = system.file("extdata/enrichr.js", package="ezRun", mustWork=TRUE)
invisible(file.copy(from=jsFile, to=basename(jsFile), overwrite=TRUE))
cat(paste0("<SCRIPT language=\"JavaScript\" SRC=\"", basename(jsFile), "\"></SCRIPT>"))
```

## Clustering results of merged samples {.tabset}

### Cell clustering without correction

Previous to cluster the cells, we performed a log-normalization of each dataset using a size factor of 10,000 molecules for each cell. In each dataset, we next aimed to identify a subset of genes with high variability across cells (i.e, they are highly expressed in some cells, and lowly expressed in others), and therefore represent heterogeneous features to prioritize for downstream analysis. We next standardized expression values for each gene across all cells in order to avoid the results to be dominated by the signal of a few highly-expressed genes. Next, we performed PCA on the scaled data using the previously determined variable features. Taking as a distance metric the previously identified PCs, the cells clusters were then identified using a graph-based clustering approach where the cells are embedded in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘communities’. 
The TSNEs below place similar cells together in low-dimensional space. The left TSNE represents cells according to the sample they belong to. On the right side, the TSNE shows the graph-based clusters determined before. 
<br/><br/>

```{r tSNE no correction plot, fig.width=8, fig.height=5}
DimPlot(scData_noCorrected, reduction = "tsne", group.by = "Plate", pt.size = 0.8)
DimPlot(scData_noCorrected, reduction = "tsne", label = TRUE, pt.size = 0.7, label.size = 5, repel = TRUE) 
```

<br/><br/>

##### The number of cells in each cluster and sample is represented in this barplot. It helps to visualize if there are existing biases when analyzing multiple datasets together (i.e. some clusters are mainly composed of cells from only one sample).

<br/>

```{r number of cells in each cluster, fig.align='center', fig.width=10, fig.height=6}
cellIdents_perSample = FetchData(object = scData_noCorrected, vars = c('ident', 'Plate'))
barplot = ggplot(data=cellIdents_perSample, aes(x=ident, fill=Plate)) + geom_bar(stat="Count")
barplot + labs(x="Cluster", y = "Number of cells", fill = "Sample")
```


```{r tSNE after integration, results='asis', fig.width=8, fig.height=5, eval = param$batchCorrection}
cat("### Cell clustering after the integration")
cat("\n\n")
cat("In addition to the previous steps that were carried out, we performed an unsupervised identification of \"anchors\" between pairs of datasets.  These anchors represent two cells (with one cell from each dataset), that we predict to originate from a common biological state. We initiate this process through dimensional reduction using canonical correlation analysis (CCA), aiming to place datasets in a shared low-dimensional space. Following dimensional reduction, we identified the K-nearest neighbors (KNNs) for each cell within its paired dataset, based on the L2-normalized CCV. Finally, we identify mutual nearest neighbors (MNN; pairs of cells, with one from each dataset, that are contained within each other’s neighborhoods). We refer to these pairwise correspondences as \"anchors\". Filtering and scoring anchors steps were performed to mitigate the effects of any incorrectly identified pairs of cells.")
cat("\n")
cat("The TSNEs below place similar cells together in low-dimensional space after the batch correction of the datasets. The left TSNE represents cells according to the sample they belong to and on the right side, the TSNE shows the graph-based clusters determined after performing the correction.")
cat("\n\n")
DimPlot(scData, reduction = "tsne", group.by = "Plate", pt.size = 0.8)
DimPlot(scData, reduction = "tsne", label = TRUE, pt.size = 0.7, label.size = 5, repel = TRUE) 
```

```{r number of cells in each cluster after correction, eval = param$batchCorrection, fig.width=10, fig.height=6, fig.align='center', results='asis'}
cat("\n")
cat("##### The number of cells in each cluster and sample after the batch correction is represented on this barplot.") 
cat("\n\n")
cellIdents_perSample = FetchData(object = scData, vars = c('ident', 'Plate'))
barplot = ggplot(data=cellIdents_perSample, aes(x=ident, fill=Plate)) + geom_bar(stat="Count")
barplot + labs(x="Cluster", y = "Number of cells", fill = "Sample")
```

### Control of the cell clustering

The plots show commonly used QC metrics that allow to explore and filter the cells. 

1. The number of unique genes detected in each cell (nGene). Low-quality cells or empty droplets will often have very few genes. Cell doublets or multiplets may exhibit a very high gene count.
2. The total number of molecules detected within a cell (nUMI). It should correlate strongly with unique genes.
3. The percentage of reads that map to the mitochondrial genome (perc_mito). Low-quality / dying cells often exhibit extensive mitochondrial contamination.

```{r}
showCellCycle = FALSE
if(!is.null(scData@meta.data$CellCycle) && !any(is.na(scData@meta.data$CellCycle)))
  showCellCycle = TRUE
```

```{r show cell cycle text, eval=showCellCycle, results='asis'}
cat("Additionally, we can visualize the cell cycle of the cells on the TSNE plot. It may be useful to identify cells with similar expression patterns and that are in the same state.")
```
<br/><br/>

```{r control cell clustering}
qc_features = c("detected", "sum", "subsets_Mito_percent")
if(!identical(setdiff(qc_features, colnames(scData@meta.data)), character(0)))
    qc_features = c("nGene", "nUMI", "perc_mito")

FeaturePlot(object = scData, features = qc_features[1], reduction = "tsne", cols = c("yellow", "red"))
FeaturePlot(object = scData, features = qc_features[2], reduction = "tsne", cols = c("yellow", "red"))
FeaturePlot(object = scData, features = qc_features[3], reduction = "tsne", cols = c("yellow", "red"))

if(showCellCycle)
  DimPlot(scData, reduction = "tsne", group.by = "CellCycle", pt.size = 0.7, plot.title="Cell cycle phase")
```

### Top cluster markers

We found positive markers that defined clusters compared to all other cells via differential expression. The test we used was the Wilcoxon Rank Sum test. Genes with an average, at least 0.25-fold difference (log-scale) between the cells in the tested cluster and the rest of the cells and an adjusted p-value < 0.05 were declared as significant. 
<br/>

```{r enrichr markers all, echo=FALSE, eval= doEnrichr(param), results='asis'}
posMarkers = scData@misc$posMarkers 

  genesPerCluster <- split(posMarkers$gene, posMarkers$cluster)
  jsCall = paste0('enrich({list: "', sapply(genesPerCluster, paste, collapse="\\n"), '", popup: true});')
  enrichrCalls <- paste0("<a href='javascript:void(0)' onClick='", jsCall, 
                         "'>Analyse at Enrichr website</a>")
  enrichrTable <- tibble(Cluster=names(genesPerCluster),
                         "# of posMarkers"=lengths(genesPerCluster),
                         "Enrichr link"=enrichrCalls)
  kable(enrichrTable, format="html", escape=FALSE,
        caption=paste0("GeneSet enrichment: genes with pvalue ", pvalue_allMarkers)) %>%
    kable_styling("striped", full_width = F, position = "left")
```

```{r markers table all, echo=FALSE}
caption ="Expression differences of cluster marker genes"
ezInteractiveTableRmd(scData@misc$posMarkers, digits=4, title=caption)
```

```{r enrichr all2all markers, eval = doEnrichr(param) && param$all2allMarkers, echo=FALSE, results='asis'}
 scData@misc$all2allMarkers = all2allMarkers  
 
 for(comparison in names(all2allMarkers)){
    write_tsv(as_tibble(all2allMarkers[[comparison]], rownames="Gene"),
              path=paste0(comparison, ".tsv"))
  }
  genesAllPerCluster <- lapply(all2allMarkers, rownames)
  genesUpPerCluster <- lapply(all2allMarkers, function(x){rownames(x)[x$avg_logFC > 0]})
  genesDownPerCluster <- lapply(all2allMarkers, function(x){rownames(x)[x$avg_logFC < 0]})
  
  jsCall_all = paste0('enrich({list: "', sapply(genesAllPerCluster, paste, collapse="\\n"), '", popup: true});')
  jsCall_up = paste0('enrich({list: "', sapply(genesUpPerCluster, paste, collapse="\\n"), '", popup: true});')
  jsCall_down = paste0('enrich({list: "', sapply(genesDownPerCluster, paste, collapse="\\n"), '", popup: true});')
  
  enrichrCalls_all <- paste0("<a href='javascript:void(0)' onClick='", jsCall_all,
                         "'>Analyse at Enrichr website</a>")
  enrichrCalls_up <- paste0("<a href='javascript:void(0)' onClick='", jsCall_up,
                         "'>Analyse at Enrichr website</a>")
  enrichrCalls_down <- paste0("<a href='javascript:void(0)' onClick='", jsCall_down,
                         "'>Analyse at Enrichr website</a>")
  enrichrTable <- tibble(Comparison=names(all2allMarkers),
                         "# of differentially expressed genes"=lengths(genesAllPerCluster),
                         "Enrichr link: all significant genes"=enrichrCalls_all,
                         "Enrichr link: up-regulated genes"=enrichrCalls_up,
                         "Enrichr link: down-regulated genes"=enrichrCalls_down,
                         "List of differentially expressed genes"=text_spec(paste0(names(all2allMarkers), ".tsv"), link=paste0(names(all2allMarkers), ".tsv")))
  kable(enrichrTable, format="html", escape=FALSE,
        caption=paste0("GeneSet enrichment: genes with pvalue ", pvalue_all2allMarkers)) %>%
    kable_styling("striped", full_width = F, position = "left")
```

### Visualization of cluster markers
#### The strongest `r param$markersToShow` gene markers (highest logFC) for each cluster are represented on the Ridge and the TSNE plots.
<br/>

```{r top cluster markers, fig.width=4, fig.height=3, results = 'asis'}
posMarkers = scData@misc$posMarkers
top_FC <- posMarkers %>% group_by(cluster) %>% top_n(param$markersToShow, avg_logFC)
for(cluster in levels(top_FC$cluster)){
  top_FC_cluster = top_FC[top_FC$cluster == cluster,]
  cat("\n")
  cat(paste0("##### Top "), param$markersToShow, ("markers for cluster "), cluster)
  cat("\n")
   for(i in 1:nrow(top_FC_cluster)){
      plot(RidgePlot(object = scData, feature = top_FC_cluster$gene[i])+ NoLegend())
      plot(FeaturePlot(object = scData, features = top_FC_cluster$gene[i], reduction = "tsne", cols = c("grey", "red")))
      
   }
   cat("\n\n")
}
```

### Cells annotation

The most challenging task in scRNA-seq data analysis is the interpretation of the results. Once we have obtained a set of clusters we want to determine what biological state is represented by each of them.
We do this using sets of marker genes that are highly expressed in each cell. This does not require matching of individual cells to the expression values of the reference dataset, which is faster and more convenient when only the identities of the markers are available. In this case, we use sets of gene markers for individual cell types taken from the CellMarkers database which contains an accurate resource of cell markers for various cell types in tissues of human and mouse (Zhang X., Lan Y., Xu J., Quan F., Zhao E., Deng C., et al. (2019). CellMarker: a manually curated resource of cell markers in human and mouse. Nucleic Acids Res. 47, D721–d728. 10.1093/nar/gky900). 
We use the AUCell package (Aibar et al. (2017) SCENIC: single-cell regulatory network inference and clustering. Nature Methods. doi: 10.1038/nmeth.4463) to identify marker sets that are highly expressed in each cell. AUCell uses the “Area Under the Curve” (AUC) to calculate whether a critical subset of the input gene set is enriched within the expressed genes for each cell. The AUC estimates the proportion of genes in the gene-set that are highly expressed in each cell. Cells expressing many genes from the gene-set will have higher AUC values than cells expressing fewer. Finally, it assigns cell type identity to each cell in the test dataset by taking the marker set with the top AUC as the label for that cell.


```{r assigments, eval= (param$species != "other"), results='asis'}
cat("We can explore the cell assignment results using different plots. Below, we show a heatmap that represents the number of cells (in log scale) from each cluster that were assigned to the different cell types. After calculating an AUC score for each cell and cell type, we assign cell type identity by taking the cell type with the top AUC as the label for that cell. Some cell types may be missing because no cells obtained their top AUC score for it.")
cat("\n\n")
```
<br/>

```{r assigments heatmap, eval= (param$species != "other"), results='asis',fig.align='center', fig.width=10, fig.height=6}
cells_AUC <- scData@misc$cells_AUC
tras_cells_AUC <- t(assay(cells_AUC))
full.labels <- colnames(tras_cells_AUC)[max.col(tras_cells_AUC)]
tab <- table(full.labels, Idents(scData))
print(pheatmap(log10(tab+10), color=viridis::viridis(100), silent=TRUE))
```

```{r tsne and AUC hist, eval= (param$species != "other"), results='asis'}
cat("The plots below show for every cell type:\n")
cat('\n')
cat("1) The distribution of the AUC values in the cells. The ideal situation will be a bi-modal distribution, in which most cells in the dataset have a low “AUC” compared to a population of cells with a higher value.  The size of the gene-set will also affect the results. With smaller gene-genes (fewer genes), it is more likely to get cells with AUC = 0. While this is the case of the “perfect markers” it is also easier to get it by chance with small datasets. The vertical bars correspond to several thresholds that could be used to consider a gene-set ‘active’. The thickest vertical line indicates the threshold selected by default: the highest value to reduce the false positives.\n")
cat('\n')
cat("2) The t-SNE can be colored based on the AUC scores. To highlight the cluster of cells that are more likely of the cell type according to the signatures, we split the cells into cells that passed the assignment threshold (colored in blue), and cells that didn’t (colored in gray).\n")
cat('\n')
cat("3) The last TSNE represents the AUC scores values. The darker a cell is the higher AUC score it obtained, i.e. the cell is more enriched in that cell type.")
cat('\n')
cellsTsne <- scData@reductions$tsne@cell.embeddings
expressionMatrix <- GetAssayData(scData, slot = "counts")
```

```{r, results='asis', fig.height=5, fig.width=5}
# graphics.off()
# par("mar")
# par(mfrow=c(5,3)) 
filtered_cells_AUC <- cells_AUC[,colSums(assay(cells_AUC))>0]
AUCell_plotTSNE(tSNE=cellsTsne, exprMat=expressionMatrix, cellsAUC=cells_AUC)
```

### Interactive explorer

<br>

[iSEE explorer](`r paste0("http://fgcz-shiny.uzh.ch/fgcz_iSEE/?data=",param$resultDir,"/SCReportMerging/sce_iSEE.rds")`)


### Data availability

##### Number and proportions of cells in each cluster and sample

[`r scData@misc$cellsPropPerClusterAndSampleFn`](`r scData@misc$cellsPropPerClusterAndSampleFn`)

##### Mean expression of every gene across the cells in each cluster

[`r scData@misc$geneMeanPerClusterFn`](`r scData@misc$geneMeanPerClusterFn`)

##### Mean expression of every gene across all the cells

[`r scData@misc$geneMeansFn`](`r scData@misc$geneMeansFn`)

##### Coordinates of every cell on the TSNE

[`r scData@misc$tSNEFn`](`r scData@misc$tSNEFn`)

##### The final Seurat object after merging is [here](scData_ObjectList.rds)

### Parameters
```{r report parameters, echo=FALSE}
param[c("npcs", "resolution", "batchCorrection","chosenClusters")]
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
