---
title:  "SCReport of merged samples"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
editor_options: 
  chunk_output_type: console
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, include=FALSE}
require(SummarizedExperiment)
require(ezRun)
require(cowplot)
require(tibble)
require(dplyr)
require(readr)
require(Seurat)
require(kableExtra)
# input for this report: sce1, sce2, param
# ans <- readRDS("/scratch/gtan/dev/SCReportMerging-p2529/FAPS4_all_SCReportMerging/ans.rds")
# sce1 <- ans$sce1
# sce2 <- ans$sce2
# param <- ans$param

sce1URL <- "https://fgcz-gstore.uzh.ch/projects/p2529/SCReport_32851_FAPS4_2019-01-08--13-45-45/FAPS4_all_SCReport/00index.html"
sce2URL <- "https://fgcz-gstore.uzh.ch/projects/p2529/SCReport_32850_2019-01-08--11-03-03/FAPS30_all_E1_SCReport/00index.html"
# param <- list()
# param$chosenClusters1 <- c()
# param$chosenClusters2 <- as.character(c(0,1,2,3,4,5,6,8,9))
# param$resolution <- 0.3

sce1 <- readRDS(file.path("/srv/gstore",
                          sub("https://fgcz-gstore.uzh.ch/", "", dirname(sce1URL), 
                              fixed=TRUE), "sce.rds"))
sce2 <- readRDS(file.path("/srv/gstore",
                          sub("https://fgcz-gstore.uzh.ch/", "", dirname(sce2URL), 
                              fixed=TRUE), "sce.rds"))
param <- metadata(sce1)$param
param$chosenClusters1 <- c()
param$chosenClusters2 <- as.character(c(0,1,2,3,4,5,6,8,9))
param$resolution <- 0.3
param$batchCorrection <- "None"

if(ezIsSpecified(param$chosenClusters1)){
  chosenCells1 <- names(metadata(sce1)$scData@ident)[metadata(sce1)$scData@ident %in% 
                                                       param$chosenClusters1]
  sce1 <- sce1[, chosenCells1]
}
if(ezIsSpecified(param$chosenClusters2)){
  chosenCells2 <- names(metadata(sce2)$scData@ident)[metadata(sce2)$scData@ident %in% 
                                                       param$chosenClusters2]
  sce2 <- sce2[, chosenCells2]
}
scData1 <- metadata(sce1)$scData
scData2 <- metadata(sce2)$scData
  
```

```{r javascript, echo=FALSE, results='asis'}
jsFile = system.file("extdata/enrichr.js", package="ezRun", mustWork=TRUE)
invisible(file.copy(from=jsFile, to=basename(jsFile), overwrite=TRUE))
cat(paste0("<SCRIPT language=\"JavaScript\" SRC=\"", basename(jsFile), "\"></SCRIPT>"))
```

## Clustering results of merged samples {.tabset}

### Cell clustering without correction

```{r tSNE no correction, include=FALSE}
sce_cbind <- S4Vectors::cbind(sce1, sce2)
metadata(sce_cbind)$param$resolution <- param$resolution
sce_cbind <- seuratPreProcess(sce_cbind)
scData <- metadata(sce_cbind)$scData

# This is to avoid error in SplitDotPlotGG
# It uses _ to separate the ident
scData@meta.data$Plate <- sub("_.*$", "", scData@meta.data$Plate)
```

```{tSNE no correction plot, echo=FALSE, fig.width=10}
p1 <- TSNEPlot(object=scData, group.by="Plate",
               do.return = TRUE, do.label = TRUE)
p1
p2 <- TSNEPlot(object=scData,
               do.return = TRUE, do.label = TRUE)
p2
```

### Cell clustering with CCA correction

```{r CCA, echo=FALSE}
  
  g.1 <- head(rownames(scData1@hvg.info), 1000)
  g.2 <- head(rownames(scData2@hvg.info), 1000)
  
  genes.use <- unique(c(g.1, g.2))
  genes.use <- intersect(genes.use, rownames(scData1@scale.data))
  genes.use <- intersect(genes.use, rownames(scData2@scale.data))
  scData.CCA <- RunCCA(scData1, scData2, genes.use = genes.use, num.cc = 30)
```

### Control of the cell clustering

```{r control cell clustering, echo=FALSE}
FeaturePlot(object = scData, features.plot = "nGene", 
            reduction.use = "tsne", no.legend = FALSE)
FeaturePlot(object = scData, features.plot = "nUMI", 
            reduction.use = "tsne", no.legend = FALSE)
```

### Conserved cell type markers

```{r conserved markers, echo=FALSE, warning=FALSE}
markers <- list()
for(eachCluster in levels(scData@ident)){
   markersEach <- FindConservedMarkers(scData, ident.1=eachCluster, 
                                       grouping.var="Plate",
                                       print.bar=FALSE, only.pos=TRUE)
  markers[[eachCluster]] <- as_tibble(markersEach, rownames="gene")
}
markers <- bind_rows(markers, .id="cluster")

posConservedMarkersFn <- "pos_conserved_markers.tsv"
write_tsv(markers, path=posConservedMarkersFn)
```

List of all positive conserved cell type markers can be accessed [here](`r posConservedMarkersFn`).

```{r enrichr markers, echo=FALSE, results='asis'}
if(doEnrichr(param)){
  genesPerCluster <- split(markers$gene, markers$cluster)
  jsCall = paste0('enrich({list: "', sapply(genesPerCluster, paste, collapse="\\n"), '", popup: true});')
  enrichrCalls <- paste0("<a href='javascript:void(0)' onClick='", jsCall, 
                         "'>Analyse at Enrichr website</a>")
  enrichrTable <- tibble(Cluster=names(genesPerCluster),
                         "# of markers"=lengths(genesPerCluster),
                         "Enrichr link"=enrichrCalls)
  kable(enrichrTable, format="html", escape=FALSE,
        caption=paste0("GeneSet enrichment analysis: conserved marker genes")) %>%
    kable_styling("striped", full_width = F, position = "left")
}
```


```{r markers table, echo=FALSE}
caption ="Conserved cell type markers"
ezInteractiveTableRmd(markers, title=caption)
```

### Visualisation of conserved markers
Top 30 conserved cell type markers are shown.

```{r conserved markers FeaturePlot, echo=FALSE, fig.width=4, fig.height=3}
markers.to.plot <- head(dplyr::arrange(markers, minimump_p_val) %>% 
                          dplyr::select(gene) %>% pull(), 30)
for(i in 1:length(markers.to.plot)){
  plot(VlnPlot(object = scData, markers.to.plot[i], do.return=TRUE))
  FeaturePlot(object = scData, features.plot = markers.to.plot[i],
              cols.use = c("gray", "red"),
              reduction.use = "tsne", no.legend = FALSE)
}
```

```{r visulisation conserved markers SplitDotPlotGG, echo=FALSE, fig.width=10, fig.height=0.5+length(unique(scData@ident)) * length(unique(scData@meta.data$Plate))*0.5}
sdp <- SplitDotPlotGG(scData, genes.plot = markers.to.plot,
                      cols.use = c("blue", "red"), x.lab.rot = T,
                      plot.legend = T, dot.scale = 8, do.return = T,
                      grouping.var = "Plate")
```

### Differential expressed genes
Identify differential expressed genes across conditions.




