### convert a fastq file into bam file, with additional tags;
### Mainly for pooled single cell samples
### fastqI1Fn: cell barcode
### fastqI2Fn: library barcode
fastq2bam <- function(fastqFn, refFn, bamFn, fastqI1Fn=NULL, fastqI2Fn=NULL){
  require(ShortRead)
  require(Biostrings)
  tempSamFn <- paste(Sys.getpid(), "temp.sam", sep="-")
  tempBamFn <- paste(Sys.getpid(), "temp", sep="-")
  
  ## SQ header; required by asBam
  seqlengths <- fasta.seqlengths(refFn)
  seqlengthsHeader <- paste0("@SQ\tSN:", names(seqlengths), "\tLN:", seqlengths)
  
  ## Add headers
  cat("@HD\tVN:1.5\tSO:unknown", file=tempSamFn, sep="\n", append=FALSE)
  cat(seqlengthsHeader, file=tempSamFn, sep="\n", append=TRUE)
  
  ## parse the fastq files
  f <- FastqStreamer(fastqFn, 1e6)
  on.exit(close(f), add=TRUE)
  if(!is.null(fastqI1Fn)){
    stopifnot(!is.null(fastqI2Fn))
    f1 <- FastqStreamer(fastqI1Fn, 1e6)
    f2 <- FastqStreamer(fastqI2Fn, 1e6)
    on.exit(close(f1), add=TRUE)
    on.exit(close(f2), add=TRUE)
  }
  while(length(fq <- yield(f))){
    samText <- paste(id(fq), "4", "*", "0", "0", "*", "*", "0", "0",
                     sread(fq), quality(quality(fq)), sep="\t")
    if(!is.null(fastqI1Fn)){
      fq1 <- yield(f1)
      fq2 <- yield(f2)
      samText <- paste0(samText, "\t", 
                        "BC:Z:", sread(fq2), "\t",
                        "QT:Z:", quality(quality(fq2)), "\t",
                        "CR:Z:", sread(fq1), "\t",
                        "CY:Z:", quality(quality(fq1)))
    }
    cat(samText, file=tempSamFn, append=TRUE, sep="\n")
  }
  tempBamFn <- asBam(file=tempSamFn, destination=tempBamFn, 
                     indexDestination=FALSE)
  bamFn <- sortBam(file=tempBamFn, destination=bamFn, byQname=TRUE, 
                   maxMemory=1024)
  on.exit(file.remove(tempSamFn), add=TRUE)
  on.exit(file.remove(tempBamFn), add=TRUE)
  invisible(bamFn)
}

### convert fastq files into a bam file, with read group tags
fastqs2bam <- function(fastqFns, fastq2Fns=NULL, readGroupNames=NULL,
                       bamFn, platform="illumina", mc.cores=ezThreads()){
  if(!isTRUE(isValidEnvironments("picard"))){
    setEnvironments("picard")
  }
  paired <- FALSE
  if(!is.null(fastq2Fns)){
    stopifnot(length(fastqFns) == length(fastq2Fns))
    paired <- TRUE
  }
  sampleBasenames <- sub("\\.(fastq|fq)(\\.gz){0,1}$", "",
                         basename(fastqFns))
  if(is.null(readGroupNames)){
    readGroupNames <- sampleBasenames
  }else{
    stopifnot(length(fastqFns) == length(readGroupNames))
  }
  
  cmd <- paste("java -Xmx4G -Djava.io.tmpdir=. -jar", 
               Sys.getenv("Picard_jar"), "FastqToSam",
               paste0("F1=", fastqFns)
               )
  if(isTRUE(paired)){
    cmd <- paste(cmd, paste0("F2=", fastq2Fns))
  }
  cmd <- paste(cmd, paste0("O=", sampleBasenames, ".bam"),
               paste0("SAMPLE_NAME=", readGroupNames),
               paste0("LIBRARY_NAME=", readGroupNames),
               paste0("READ_GROUP_NAME=", readGroupNames),
               paste0("RUN_DATE=", format(Sys.time(), "%Y-%m-%dT%H:%M:%S+00:00")),
               paste0("PLATFORM=", platform),
               "SEQUENCING_CENTER=FGCZ")
  
  ezMclapply(cmd, ezSystem, mc.preschedule=FALSE, mc.cores = mc.cores)
  
  ## picard tools merge
  cmd <- paste("java -Djava.io.tmpdir=. -jar", 
               Sys.getenv("Picard_jar"), "MergeSamFiles",
               "SORT_ORDER=unsorted",
               paste0("I=", paste0(sampleBasenames, ".bam"), collapse=" "),
               paste0("O=", bamFn),
               "USE_THREADING=true", "SORT_ORDER=queryname")
  ezSystem(cmd)
  file.remove(paste0(sampleBasenames, ".bam"))
  
  invisible(bamFn)
}

### Convert  bam/sam files into fastqs
bam2fastq <- function(bamFn, OUTPUT_PER_RG=TRUE, OUTPUT_DIR=".",
                      paired=FALSE,
                      fastqFns=sub("(\\.bam|\\.sam)$", "_R1.fastq", bamFn),
                      fastq2Fns=sub("(\\.bam|\\.sam)$", "_R2.fastq", bamFn)){
  setEnvironments("picard")
  stopifnot(length(bamFn) == 1L)
  
  if(isTRUE(OUTPUT_PER_RG)){
    ## I don't want to parse the bam header to get RG IDs
    ## Put them in a tempdir and move to OUTPUT_DIR later
    tempDIR <- paste("SamtoFastqTempDir", Sys.getpid(), sep="-")
    dir.create(tempDIR)
    on.exit(unlink(tempDIR, recursive=TRUE), add = TRUE)
    cmd <- paste("java -Djava.io.tmpdir=. -jar", 
                 Sys.getenv("Picard_jar"), "SamToFastq",
                 paste0("I=", bamFn),
                 paste0("OUTPUT_DIR=", tempDIR),
                 "OUTPUT_PER_RG=true RG_TAG=ID"
                 )
   ezSystem(cmd)
   fastqFns <- list.files(path=tempDIR, pattern="_1\\.fastq$")
   fromFns <- file.path(tempDIR, fastqFns)
   toFns <- file.path(OUTPUT_DIR, sub("_1\\.fastq$", "_R1.fastq", fastqFns))
   file.rename(from=fromFns, to=toFns)
   if(isTRUE(paired)){
     file.rename(from=sub("_1\\.fastq$", "_2.fastq", fromFns), 
                 to=sub("_R1\\.fastq$", "_R2.fastq", toFns)
                 )
   }
   return(invisible(toFns))
  }else{
    cmd <- paste("java -Djava.io.tmpdir=. -jar", 
                 Sys.getenv("Picard_jar"), "SamToFastq",
                 paste0("I=", bamFn),
                 paste0("FASTQ=", fastqFns))
    if(isTRUE(paired))
      cmd <- paste(cmd, paste0("SECOND_END_FASTQ=", fastq2Fns))
    ezSystem(cmd)
    return(invisible(fastqFns))
  }
}

ezMethodBam2Fastq <- function(input=NA, output=NA, param=NA,
                              OUTPUT_PER_RG=TRUE){
  require(Biostrings)
  
  if(isTRUE(OUTPUT_PER_RG)){
    output = EzDataset(file=input$getFullPaths("CellDataset"),
                       dataRoot=param$dataRoot)
    output$setColumn("Read1", paste0(getwd(), "/", output$getNames(),
                                     "_R1.fastq"))
    if (param$paired){
      output$setColumn("Read2", paste0(getwd(), "/", output$getNames(), 
                                       "_R2.fastq"))
    } else {
      if ("Read2" %in% input$colNames)
        output$setColumn("Read2", NULL)
    }
    output$dataRoot = NULL
    
    bam2fastq(bamFn=input$getFullPaths("Read1"),
              OUTPUT_PER_RG=TRUE, OUTPUT_DIR=getwd(),
              paired=param$paired)
  }else{
    if (!is(output, "EzDataset")){
      output = input$copy()
      output$setColumn("Read1", paste0(getwd(), "/", input$getNames(), 
                                       "-R1.fastq"))
      if (param$paired){
        output$setColumn("Read2", paste0(getwd(), "/", input$getNames(), 
                                         "-R2.fastq"))
      } else {
        if ("Read2" %in% input$colNames){
          output$setColumn("Read2", NULL)
        }
      }
      output$dataRoot = NULL
    }
    bam2fastq(bamFn=input$getFullPaths("Read1"),
              OUTPUT_PER_RG=FALSE,
              fastqFns=output$getColumn("Read1"),
              fastq2Fns=ifelse(isTRUE(param$paired), output$getColumn("Read2"),
                               NULL),
              paired=param$paired)
    output$setColumn("Read Count",
                     sapply(output$getColumn("Read1"), fastq.geometry)[1, ])
  }
  return(output)
}
